pkg/util/spanlogger/spanlogger.go

```diff
@@ -94,3 +95,22 @@ func (s *SpanLogger) Error(err error) error {
 	s.Span.LogFields(otlog.Error(err))
 	return err
 }
+
+// triggerError calls s.Error if SpanLogger's log level is set to level.ErrorValue().
+func (s *SpanLogger) triggerError(kvps ...interface{}) {
```

Comment:
I think this function is not quite correct. Behaviour I would expect to see is:

1) if log level is error, set span to error (`ext.Error.Set(s.Span, true)`). That is all that `triggerError` should do imho.

2) if there is also error object logged, it should be converted to `otlog.Error(err)` field. But I would expect to see this happen in `Log` method, specifically in `InterleavedKVToFields` mapping, which returns fields from values. Note that presence of error object should not mark span itself as errorneous, as errors an be logged at debug level too.

WDYT?

Reply:
I see what you mean: 
Right now `triggerError` calls `SpanLogger.Error`, which calls `ext.Error.Set(s.Span, true)` and `s.Span.LogFields(otlog.Error(err))`. `err` is being logged twice, once in the triggered `SpanLogger.Error` and once in `Log` itself.

 #3016 wants to save users one line of code (calling `SpanLogger.Error` themselves)  if `level.Error` is called before `SpanLogger.Log` - `triggerError` simply allows user to achieve what the two lines of code below do in one line.
https://github.com/cortexproject/cortex/blob/8fe0bb18f27189fcf7281099e9234d2b25ad5f53/pkg/testexporter/correctness/simple.go#L179-L180
 
 Please let me know if anything above is unclear.

Reply:
@pstibrany I wonder if you have any further thoughts?

Reply:
> Right now `triggerError` calls `SpanLogger.Error`, which calls `ext.Error.Set(s.Span, true)` and `s.Span.LogFields(otlog.Error(err))`. `err` is being logged twice, once in the triggered `SpanLogger.Error` and once in `Log` itself.

You are right, that's not good :)

I see also that we cannot easily modify `InterleavedKVToFields`. What about following approach:
- `Log` will first call `s.Logger.Log(kvps...)` as it does now
- it would then have logic from `triggerError`, specifically
    - lookup of `level.Key()=level.ErrorValue()` pair (this causes `ext.Error.Set(s.Span, true)` to be called)
    - if we have found `level.Key()=level.ErrorValue()`, then we would also search for non-nil objects implementing `error` interface. If found with `err` or `error` key, we can remove it from `kvps` slice, and keep as extra field. (Only first found error is handled this way).
- `Log` would then call `otlog.InterleavedKVToFields` to convert remaining fields, and if we kept extra error field in previous step, we add it to these fields
- `Log` then calls `s.Span.LogFields` with all fields (incl. error one, if any).

We would not reuse `(*SpanLogger).Error` method, to avoid logging of error twice.

WDYT?

Reply:
May I ask the motivation to take out the first `error`at step 2, and putting it back at step 3? Do you mean that the first error should be kept, and the rest of the `error` should be removed from the `kvps`?

Reply:
> May I ask the motivation to take out the first `error`at step 2, and putting it back at step 3?

Motivation is to avoid double-logging of the error, and to store it wrapped into `otlog.Error(err)` instead. Calling `otlog.InterleavedKVToFields` with `error` instance would convert it to `string`.

> Do you mean that the first error should be kept as an extra field,

Yes.

> and the rest of the `error` should be removed from the `kvps`?

No, I would keep the rest as-is. If we removed them, they would disappear from the span-log. That would be bad.